SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/ReferenceCountedLockingPolicy.cpp

```diff
@@ -163,3 +177,136 @@ bool UReferenceCountedLockingPolicy::ReleaseLockFromDelegate(AActor* ActorToRele
+
+	check(LockedOwnershipPathActorToExplicitlyLockedActors.Contains(DestroyedOwnershipPathActor));
+
+	TArray<const AActor*> ExplicitlyLockedActorsWithThisActorInOwnershipPath = LockedOwnershipPathActorToExplicitlyLockedActors.FindAndRemoveChecked(DestroyedOwnershipPathActor);
+
+	for (const AActor* ExplicitlyLockedActor : ExplicitlyLockedActorsWithThisActorInOwnershipPath)
+	{
+		check(ActorToLockingState.Contains(ExplicitlyLockedActor));
+		ResetLockedActorOwnershipHierarchyInformation(ExplicitlyLockedActor);
+	}
+
+	check(!LockedOwnershipPathActorToExplicitlyLockedActors.Contains(DestroyedOwnershipPathActor));
+}
+
+void UReferenceCountedLockingPolicy::ResetLockedActorOwnershipHierarchyInformation(const AActor* ExplicitlyLockedActor, const AActor* DeletedHierarchyActor)
+{
+	//	Get the old ownership path :
+	//		Update hierarchy root mapping to remove this Actor
+	//		Update ownership path Actor mapping to remove this Actor
+	//	Set new Actor ownership path in mapping
+	//	Get the new ownership hierarchy of locked Actor :
+	//		Add hierarchy root mapping to explicitly locked Actor
+	//		Add ownership path Actor mapping to explicitly locked Actor
+
+	TArray<AActor*> OldOwnershipHierarchyPath = ActorToLockingState.FindChecked(ExplicitlyLockedActor).OwnershipPath;
+	RemoveOwnershipHierarchyPathInformation(OldOwnershipHierarchyPath);
+
+	TArray<AActor*> NewOwnershipHierarchyPath = SpatialGDK::GetOwnershipHierarchyPath(ExplicitlyLockedActor, DeletedHierarchyActor);
+	ActorToLockingState.FindChecked(ExplicitlyLockedActor).OwnershipPath = NewOwnershipHierarchyPath;
+	AddOwnershipHierarchyPathInformation(ExplicitlyLockedActor, NewOwnershipHierarchyPath);
+}
+
+void UReferenceCountedLockingPolicy::RemoveOwnershipHierarchyPathInformation(TArray<AActor*> OwnershipHierarchyPath)
+{
+	for (AActor* OwnershipPathActor : OwnershipHierarchyPath)
+	{
+		// Find Actors in this root Actor's hierarchy which are explicitly locked
+		TArray<const AActor*>& ExplicitlyLockedActorsWithThisActorOnPath = LockedOwnershipPathActorToExplicitlyLockedActors.FindOrAdd(OwnershipPathActor);
+		check(ExplicitlyLockedActorsWithThisActorOnPath.Num() > 0);
+
+		// If there's only one explicitly locked Actor in the hierarchy, we're removing the only Actor with this owner,
+		// so we can stop caring about the root itself. Otherwise, just remove the specific Actor entry it the root's list.
+		if (ExplicitlyLockedActorsWithThisActorOnPath.Num() == 1)
+		{
+			LockedOwnershipPathActorToExplicitlyLockedActors.Remove(OwnershipPathActor);
+			OwnershipPathActor->OnDestroyed.RemoveDynamic(this, &UReferenceCountedLockingPolicy::OnOwnershipPathActorDeleted);
+		}
+		else
+		{
+			ExplicitlyLockedActorsWithThisActorOnPath.Remove(OwnershipPathActor);
+		}
+	}
+}
+
+void UReferenceCountedLockingPolicy::AddOwnershipHierarchyPathInformation(const AActor* ExplicitlyLockedActor, TArray<AActor*> OwnershipHierarchyPath)
```

Comment:
It doesn't look like this is taking ownership over a TArray so it should be a const &

Reply:
Or just a reference if you need to modify the actors inside it.